lib/Sema/BuilderTransform.cpp

```diff
@@ -1853,33 +1853,66 @@ std::vector<ReturnStmt *> TypeChecker::findReturnStatements(AnyFunctionRef fn) {
 
 bool TypeChecker::typeSupportsBuilderOp(
     Type builderType, DeclContext *dc, Identifier fnName,
-    ArrayRef<Identifier> argLabels, SmallVectorImpl<ValueDecl *> *allResults) {
+    ArrayRef<Identifier> argLabels,
+    SmallVectorImpl<BuilderOpMismatch> *nearMisses) {
+
+  auto recordNearMiss = [&](ValueDecl *vd, BuilderOpMismatchKind kind) {
+    if (nearMisses)
+      nearMisses->emplace_back(vd, kind);
+  };
+
   bool foundMatch = false;
   SmallVector<ValueDecl *, 4> foundDecls;
   dc->lookupQualified(
       builderType, DeclNameRef(fnName),
       NL_QualifiedDefault | NL_ProtocolMembers, foundDecls);
   for (auto decl : foundDecls) {
-    if (auto func = dyn_cast<FuncDecl>(decl)) {
-      // Function must be static.
-      if (!func->isStatic())
-        continue;
+    if (isa<EnumElementDecl>(decl)) {
+      recordNearMiss(decl, BuilderOpMismatchKind::EnumCase);
+      continue;
+    } else if (!isa<FuncDecl>(decl)) {
+      recordNearMiss(decl, BuilderOpMismatchKind::Other);
+      continue;
+    }
 
-      // Function must have the right argument labels, if provided.
-      if (!argLabels.empty()) {
-        auto funcLabels = func->getName().getArgumentNames();
-        if (argLabels.size() > funcLabels.size() ||
-            funcLabels.slice(0, argLabels.size()) != argLabels)
-          continue;
+    auto *func = cast<FuncDecl>(decl);
+    // Function must be static.
+    if (!func->isStatic()) {
+      recordNearMiss(func, BuilderOpMismatchKind::InstanceMethod);
+      continue;
+    }
+
+    // Function must have the right argument labels, if provided.
+    if (!argLabels.empty()) {
+      auto funcLabels = func->getName().getArgumentNames();
+      if (argLabels.size() > funcLabels.size() ||
+          funcLabels.slice(0, argLabels.size()) != argLabels) {
```

Comment:
This is an approximation of the rule used to actually match call arguments. Do you want to go all the way and use `matchCallArguments` to get it exactly right?

Reply:
I think that makes sense for most of the `build*` functions (since the transform introduces a fixed number of arguments), but I don't think it would work for `buildBlock` since we won't know the correct number of arguments until we're actually evaluating a call, and every call to `buildBlock` may differ in the number of arguments.

It seems to me that, given that `buildBlock` calls may have an unbounded number of arguments (of any type), the best we can do when evaluating the `buildBlock` declaration is make sure that there are no required labeled arguments. If I'm missing something let me know!